### PR TITLE
Change stream cache eviction to least-recently-used

### PR DESCRIPTION
**Pull request by leviroth**
This is the same patch as https://github.com/praw-dev/praw/pull/1547, but applied to the equivalent files in Async PRAW.

I haven't made any effort to test beyond importing the unit test patch, nor do I know anything about Async PRAW beyond some general knowledge of asyncio.

### CODE REVIEW

File: tests/unit/models/test_util.py [hunk 1]
```diff
@@ -38,7 +45,58 @@ def test_exponential_counter__reset(self):
...
+        [bset.add(i) for i in range(10)]
+        assert 0 in bset
+        bset.add(10)
+        assert 0 in bset
+        assert 1 not in bset
+
+
+class TestStream(UnitTest):
+    @mock.patch("time.sleep", return_value=None)
+    def test_stream(self, _):
```
Comment (PythonCoderAS) [inline tests/unit/models/test_util.py:79]:
```suggestion
    async def test_stream(self, _):
```

File: tests/unit/models/test_util.py [hunk 2]
```diff
@@ -38,7 +45,58 @@ def test_exponential_counter__reset(self):
...
+        Thing = namedtuple("Thing", ["fullname"])
+        initial_things = [Thing(n) for n in reversed(range(100))]
+        counter = 99
+
+        def generate(limit, **kwargs):
```
Comment (PythonCoderAS) [inline tests/unit/models/test_util.py:84]:
```suggestion
        async def generate(limit, **kwargs):
```

File: tests/unit/models/test_util.py [hunk 3]
```diff
@@ -38,7 +45,58 @@ def test_exponential_counter__reset(self):
...
+            if counter % 2 == 0:
+                return initial_things
+            return [Thing(counter)] + initial_things[:-1]
+
+        stream = stream_generator(generate)
+        seen = set()
+        for _ in range(400):
+            thing = next(stream)
+            assert thing not in seen
+            seen.add(thing)
```
Comment (PythonCoderAS) [inline tests/unit/models/test_util.py:96]:
```suggestion
        loop_counter = 0
        async for thing in stream:
            assert thing not in seen
            seen.add(thing)
            counter += 1
            if counter == 400:
                break
```

**Review by PythonCoderAS — CHANGES_REQUESTED** [review]
A few changes to be made in order to make the test async-compliant. I had to use an external counter since I am not aware of a next() function for async generators.

**Comment by PythonCoderAS** [comment]
Upon further review, it seems that a custom async iterator is going to have to replace the function.

**Comment by LilSpazJoekp** [comment]
> Upon further review, it seems that a custom async iterator is going to have to replace the function.

I have helper functions made here: https://github.com/praw-dev/asyncpraw/blob/ad443d6a9218e8868dfcd33d53ae9fd3508c9ce7/tests/integration/__init__.py#L76-L85

Note that `async_next()` is not a true implementation of `next()`. I think using `async_list()` to convert it to a list would be best as you can then you can use `next()`.

**Comment by PythonCoderAS** [comment]
I think the issue here is that in normal PRAW, the function returns a normal iterator (a list). The function therefore should logically return an async iterator. The best idea would be an async list, which inherits from list, and the `__anext__` method should yield the list's attributes back.

**Comment by leviroth** [comment]
As a heads up, I'm not planning to work on this patch any further. Feel free to do whatever you like with it.

**Pull request merged**